File: OneCodeTeam/Socket-programming-in-Silverlight-19244/[VB]-Socket programming in Silverlight (VBSL4SocketProgramming)/VB/VBSL4SocketProgramming/UdpSingleSourceMulticastChannel.vb
'**************************** Module Header ******************************\
' Module Name: UdpSingleSourceMulticastChannel.vb
' Project:     VBSL4SocketProgramming
' Copyright (c) Microsoft Corporation
'
' The project illustrates how to use Socket achieve MultiCast function
' in Silverlight. We demonstrate ISM and SSM in this sample, The 
' application includes a console application as server side and a 
' Silverlight application as client side, the server will broadcast
' messages to client sides and client sides can also send messages 
' to other users.
'
' This class use to create a SSM client.
' 
' This source is subject to the Microsoft Public License.
' See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
' All other rights reserved.
' 
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
' WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
'****************************************************************************




Imports System.Net.Sockets

Public Class UdpSingleSourceMulticastChannel
    Implements IDisposable
    Private SSMClient As UdpSingleSourceMulticastClient
    Private byteBuffer As Byte()
    Private serverIP As IPAddress
    Private targetIP As IPAddress
    Private port As Integer
    Public IsJoinBroadcast As Boolean
    Public IsReceived As Boolean
    Public Event AfterOpen As EventHandler
    Public Event BeforeReceived As EventHandler(Of UdpConvertMessageEvent)
    Public Event BeforeCloseSSM As EventHandler

    ''' <summary>
    ''' Constructor method. Define a buffer bytes variable,
    ''' server IP, group IP, port, and a SSM client.
    ''' </summary>
    ''' <param name="serverIP"></param>
    ''' <param name="targetIP"></param>
    ''' <param name="port"></param>
    ''' <param name="messageSize"></param>
    ''' <remarks></remarks>
    Public Sub New(ByVal serverIP As IPAddress, ByVal targetIP As IPAddress, ByVal port As Integer, ByVal messageSize As Integer)
        Me.serverIP = serverIP
        Me.targetIP = targetIP
        Me.port = port
        byteBuffer = New Byte(messageSize - 1) {}
        SSMClient = New UdpSingleSourceMulticastClient(serverIP, targetIP, port)
    End Sub

    ''' <summary>
    ''' Connect to SSM server method.
    ''' </summary>
    ''' <remarks></remarks>
    Public Sub Open()
        If Me.isJoinBroadcast Then
            Return
        End If
        If SSMClient Is Nothing Then
            SSMClient = New UdpSingleSourceMulticastClient(serverIP, targetIP, port)
        End If
        Dim callBack As New AsyncCallback(AddressOf OpenCallBack)
        SSMClient.BeginJoinGroup(callBack, Nothing)
    End Sub

    Private Sub OpenCallBack(ByVal result As IAsyncResult)
        SSMClient.EndJoinGroup(result)
        Me.isJoinBroadcast = True
        Me.isReceived = True
        Deployment.Current.Dispatcher.BeginInvoke(AddressOf OpenInvoke)
    End Sub

    Private Sub OpenInvoke()
        RaiseEvent AfterOpen(Me, EventArgs.Empty)
        Me.Received()
    End Sub

    ''' <summary>
    ''' Receive message from ASM server method.
    ''' </summary>
    ''' <remarks></remarks>
    Public Sub Received()
        If Not Me.isJoinBroadcast Then
            Return
        End If
        If Me.isReceived Then
        Array.Clear(byteBuffer, 0, byteBuffer.Length)
            SSMClient.BeginReceiveFromSource(byteBuffer, 0, byteBuffer.Length,
                 Function(result)
                     If Me.isJoinBroadcast Then
                         SSMClient.EndReceiveFromSource(result, sourcePoint)
                         Deployment.Current.Dispatcher.BeginInvoke(
                             Function()
                                 If Me.isJoinBroadcast Then
                                     RaiseEvent BeforeReceived(Me, New UdpConvertMessageEvent(byteBuffer, New IPEndPoint(serverIP, sourcePoint)))
                                     Me.Received()
                                     Return vbEmpty
                                 Else
                                     Return vbEmpty
                                 End If
                             End Function
                         )
                         Return vbEmpty
                     Else
                         Return vbEmpty
                     End If
                 End Function, Nothing)
        End If
    End Sub
    Dim sourcePoint As Integer
    ''' <summary>
    ''' Close the connection event.
    ''' </summary>
    ''' <remarks></remarks>
    Public Sub Close()
        Me.isJoinBroadcast = False
        RaiseEvent BeforeCloseSSM(Me, EventArgs.Empty)
        Dispose(True)
    End Sub
    ' To detect redundant calls
    Private disposedValue As Boolean = False

    Protected Overridable Sub Dispose(ByVal disposing As Boolean)
        If Not Me.disposedValue Then
            If disposing Then
                If SSMClient IsNot Nothing Then
                    SSMClient.Dispose()
                    SSMClient = Nothing
                End If
            End If
        End If
        Me.disposedValue = True
    End Sub

    Public Sub Dispose() Implements IDisposable.Dispose
        Dispose(True)
        GC.SuppressFinalize(Me)
    End Sub

End Class

File: OneCodeTeam/Socket-programming-in-Silverlight-19245/[C#]-Socket programming in Silverlight (CSSL4SocketProgramming)/C#/CSSL4SocketProgramming/UdpSingleSourceMulticastChannel.cs
/****************************** Module Header ******************************\
* Module Name: UdpSingleSourceMulticastChannel.cs
* Project:     CSSL4SocketProgramming
* Copyright (c) Microsoft Corporation
*
* The project illustrates how to use Socket achieve MultiCast function
* in Silverlight. We demonstrate ISM and SSM in this sample, The 
* application includes a console application as server side and a 
* Silverlight application as client side, the server will broadcast
* messages to client sides and client sides can also send messages 
* to other users.
*
* This class use to create a SSM client.
* 
* This source is subject to the Microsoft Public License.
* See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
* All other rights reserved.
* 
* THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
* EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
* WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
\*****************************************************************************/



using System;
using System.Net;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Documents;
using System.Windows.Ink;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;
using System.Net.Sockets;
using System.Text;

namespace CSSL4SocketProgramming
{
    public class UdpSingleSourceMulticastChannel : IDisposable
    {
        private UdpSingleSourceMulticastClient SSMClient;
        private byte[] byteBuffer;
        private IPAddress serverIP;
        private IPAddress targetIP;
        private int port;
        private int sourcePoint;
        public bool IsJoinBroadcast;
        public bool IsReceived = true;
        public event EventHandler AfterOpen;
        public event EventHandler<UdpConvertMessageEvent> BeforeReceived;
        public event EventHandler BeforeCloseSSM;

        /// <summary>
        /// Constructor method. Define a buffer bytes variable,
        /// server IP, group IP, port, and a SSM client.
        /// </summary>
        /// <param name="serverIP"></param>
        /// <param name="targetIP"></param>
        /// <param name="port"></param>
        /// <param name="messageSize"></param>
        public UdpSingleSourceMulticastChannel(IPAddress serverIP, IPAddress targetIP, int port, int messageSize)
        {
            this.serverIP = serverIP;
            this.targetIP = targetIP;
            this.port = port;
            byteBuffer = new byte[messageSize];
            SSMClient = new UdpSingleSourceMulticastClient(serverIP, targetIP, port);
        }

        /// <summary>
        /// Connect to SSM server method.
        /// </summary>
        public void Open()
        {
            if (this.IsJoinBroadcast)
            {
                return;
            }
            if (SSMClient == null)
            {
                SSMClient = new UdpSingleSourceMulticastClient(serverIP, targetIP, port);
            }
            AsyncCallback callBack = new AsyncCallback(OpenCallBack);
            SSMClient.BeginJoinGroup(callBack, null);
        }

        private void OpenCallBack(IAsyncResult result)
        {
            SSMClient.EndJoinGroup(result);
            this.IsJoinBroadcast = true;
            this.IsReceived = true;
            Deployment.Current.Dispatcher.BeginInvoke(OpenEvent);
        }

        private void OpenEvent()
        {
            var openHandler = AfterOpen;
            if (openHandler != null)
                openHandler(this, EventArgs.Empty);
            this.Received();
        }

        /// <summary>
        /// Receive message from ISM server method.
        /// </summary>
        public void Received()
        {
            if (!this.IsJoinBroadcast)
            {
                return;
            }
            if (this.IsReceived)
            {
                Array.Clear(byteBuffer, 0, byteBuffer.Length);
                SSMClient.BeginReceiveFromSource(byteBuffer, 0, byteBuffer.Length, result =>
                {
                    if (this.IsJoinBroadcast)
                    {
                        SSMClient.EndReceiveFromSource(result, out sourcePoint);
                        Deployment.Current.Dispatcher.BeginInvoke(() =>
                        {
                            if (this.IsJoinBroadcast)
                            {
                                var receivedHandler = BeforeReceived;
                                if (receivedHandler != null)
                                {
                                    receivedHandler(this, new UdpConvertMessageEvent(byteBuffer, new IPEndPoint(serverIP, sourcePoint)));
                                }
                                this.Received();
                            }
                            else
                            {
                                return;
                            }
                        });
                    }
                    else
                    { return; }
                }, null);
            }
        }

        /// <summary>
        /// Close the connection event.
        /// </summary>
        public void Close()
        {
            this.IsJoinBroadcast = false;
            var closeHandler = BeforeCloseSSM;
            if (closeHandler != null)
            {
                closeHandler(this, EventArgs.Empty);
            }
            Dispose(true);
        }

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        // To detect redundant calls
        private bool disposed = false;
        protected virtual void Dispose(bool disposing)
        {
            if (disposed)
            {
                return;
            }
            if (disposing)
            {
                if (SSMClient != null)
                {
                    SSMClient.Dispose();
                    SSMClient = null;
                }
            }
            disposed = true;
        }
    }
}
